backend/eventloop/db_reconciler.go

```diff
@@ -444,6 +469,360 @@ func cleanRowFromDB(ctx context.Context, dbQueries db.DatabaseQueries, id string
+
+	offSet := 0
+	// Continuously iterate and fetch batches until all entries of Application table are processed.
+	for {
+		if offSet != 0 {
+			time.Sleep(sleepIntervalsOfBatches)
+		}
+
+		var listOfApplicationsFromDB []db.Application
+
+		// Fetch Application table entries in batch size as configured above.​
+		if err := dbQueries.GetApplicationBatch(ctx, &listOfApplicationsFromDB, appRowBatchSize, offSet); err != nil {
+			log.Error(err, fmt.Sprintf("Error occurred in cleanOrphanedEntriesfromTable_Application while fetching batch from Offset: %d to %d: ",
+				offSet, offSet+appRowBatchSize))
+			break
+		}
+
+		// Break the loop if no entries are left in table to be processed.
+		if len(listOfApplicationsFromDB) == 0 {
+			log.Info("All Application entries are processed by cleanOrphanedEntriesfromTable_Application.")
+			break
+		}
+
+		// Iterate over batch received above.
+		for _, appDB := range listOfApplicationsFromDB {
+
+			// Check if application has entry in DTAM table, if not then delete the application
+			// If created time is less than waitTimeforRowDelete then ignore dont delete even if ACTDM entry is missing.
+			if !slices.Contains(listOfAppsIdsInDTAM[string(dbType_Application)], appDB.Application_id) &&
+				time.Since(appDB.Created_on).Hours() > waitTimeforRowDelete {
+
+				// Fetch Application from DB and convert Spec into an Object as we need namespace for creating Operations CR.
+				var appArgo fauxargocd.FauxApplication
+				isApplicationPresent := true
+				if err := yaml.Unmarshal([]byte(appDB.Spec_field), &appArgo); err != nil {
+					log.Error(err, "Error occurred in cleanOrphanedEntriesfromTable_Application while unmarshalling application: "+appDB.Application_id)
+					isApplicationPresent = false
+				}
+
+				_, err := dbQueries.DeleteApplicationById(ctx, appDB.Application_id)
+				if err != nil {
+					log.Error(err, "Error occurred in cleanOrphanedEntriesfromTable_Application while cleaning Application entry "+appDB.Application_id+" from DB.")
+					continue
+				} else {
+					log.Info("Application entry: " + appDB.Application_id + " is successfully deleted by cleanOrphanedEntriesfromTable_Application.")
+				}
+
+				// Skip if Application was not found, as we need namespace for creating Operation CR.
+				if isApplicationPresent {
+					// Create k8s Operation to delete related CRs using Cluster Agent
+					createOperation(ctx, appDB.Engine_instance_inst_id, appDB.Application_id, appArgo.Namespace, db.OperationResourceType_Application, dbQueries, client, log)
+				}
+			}
+		}
+
+		// Skip processed entries in next iteration
+		offSet += appRowBatchSize
+	}
+}
+
+// getListOfCRIdsFromTable loops through DTAMs in database and returns list of resource IDs for each CR type (i.e. RepositoryCredential, ManagedEnvironment, SyncOperation) .
+func getListOfCRIdsFromTable(ctx context.Context, dbQueries db.DatabaseQueries, tableType string, log logr.Logger) map[string][]string {
```

Comment:
I understand we are prefetching all the CR IDs to verify whether there is a corresponding entry in the DTAM/ACTDM table. But they are retrieved in batches with a sleep interval and I wonder if this would make it slower than directly checking if an individual entry exists in the table using something like `GetDeploymentToApplicationMappingByApplicationId`. WDYT?

Reply:
Yes, it can be done, but in that case there will be many DB calls, that is what we want to avoid and it is deleting entries which are older than 1 Hour, it means there is enough time for components to update DB entries for CRs and there should not be any chance of race conditions.